### PR TITLE
feat: Propagate Kahan compensation through `Potential` protocol for exact MC energy differences

### PR DESCRIPTION
**Pull request by n-gao**
Kahan compensation is now threaded through the entire potential pipeline so that Monte Carlo acceptance ratios stay numerically exact even when a move's energy change is far below the ULP of the running total.

## Motivation

In single precision, `ulp(1e5 eV) ≈ 8 meV`, which is larger than many physically relevant MC move energies (~2 meV). Naively subtracting two large cached totals to get `ΔU` loses the move entirely to rounding. The fix is to keep a Kahan compensation term alongside every cached total and difference the two compensated accumulators instead of the raw values.

### `KahanSummand` additions
- `__mul__` / `__rmul__` so that `ScaledPotential` can scale the compensation alongside the value.

### Potential protocol — `include_compensate` keyword
Every `Potential` implementation (`SummedPotential`, `ScaledPotential`, `CachedPotential`, `MappedPotential`, `PotentialFromEnergy`, `DirectPotential`, `InternalEnergies`) now accepts `include_compensate: bool = False`. When `False` (the default) the public API is unchanged — callers receive a plain `PotentialOut`. When `True` they receive a `KahanSummand[PotentialOut]` carrying the accumulated rounding error. Two new type aliases, `PotentialResult` and `CompensatedPotentialResult`, capture the two return shapes, and `@overload` signatures make the distinction visible to type checkers.

- `SummedPotential` always requests compensated outputs from its children and sums them with Kahan addition; it strips the compensation before returning unless the caller asked for it.
- `CachedPotential` gains an optional `compensate_cache` lens. When set, the compensation is committed to state alongside the cached value so the next call can read back a fully compensated accumulator. `cached_value` now returns a `KahanSummand` (compensation is zero when no `compensate_cache` is configured).

### `BoltzmannLogProbabilityRatio`
`ΔU` is now computed via `KahanSummand.difference(old, new)` rather than plain subtraction, recovering the exact energy change from two compensated accumulators.

### `MCMCSystems.potential_energy`
Changed from `Array` to `KahanSummand[Array]`. Initialised to `KahanSummand.init(jnp.zeros(1))`. All read sites use `.total`; the `make_boltzmann_probability_ratio` helper wires the `.value` and `.compensate` fields into separate cache lenses so both halves of the accumulator survive accepted moves.

### `HasCompensatedPotentialEnergy` protocol
Replaces `HasPotentialEnergy` wherever the field is a `KahanSummand`, keeping the plain protocol for contexts that still expose a raw array.

### Tests
A new `TestCompensatedEnergyDifference` test demonstrates the problem and the fix in float32: after 20 accepted moves of size 2 meV on a 100 keV baseline, the compensated log-ratio matches the exact value to within 0.01 %, while the naïve difference is off by more than 50 %.

---

## Results

Differencing two compensated totals removes the `ulp(|E|)` quantization from the acceptance energy difference, so the benefit grows with total energy. Measured per move against float64 recomputes of the same configurations:

| total energy | mean error, before | mean error, after | advantage |
|---|---|---|---|
| 9.3e3 eV | 6.99e-4 eV | 3.65e-4 eV | 1.9x, closer on 14/17 moves |
| 2.2e4 eV | 5.51e-4 eV | 3.28e-5 eV | 16.8x, closer on 11/11 moves |
| 1e5 eV (synthetic, quarter-ulp move) | >50% wrong every step | exact to rtol 1e-4 | - |

The first two rows used different equilibration and Ewald settings, so they show the trend rather than a controlled sweep; the change becomes decisive as `ulp(|E|)` approaches kT, around |E| ~ 1e5 eV. In float64 it is inert: full-length runs are bit-identical with and without it across ~3e5 acceptance decisions (suite 1895 passed / 7 skipped).

Scope: ablating the compensation entirely, with the Ewald self-interaction fix from the follow-up PR in place, changes float32 heat of adsorption by less than the 1-4% replica error at n=4 on all five frameworks of that batch (|E| = 345-1041 eV, where 1 ulp is 1e-3 to 5e-3 kT). That fix, not this change, accounts for the float32 recovery there. This is kept because it costs ~2 flops per accumulation and removes a failure mode that grows with system size.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#269** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/269?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#278** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/278?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#272** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/272?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#271** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/271?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#268** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/268?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#267** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/267?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/267?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#229** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/229?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#233** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/233?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#235** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/235?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#234** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/234?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#232** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/232?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#230** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/230?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#228** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/228?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#227** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/227?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#226** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/226?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#224** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/224?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#223** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/223?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#222** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/222?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#215** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/215?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#213** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/213?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#216](https://github.com/cusp-ai-oss/kUPS/pull/216) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/216?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#212** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/212?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#211** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/211?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#210** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/210?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#204** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/204?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#209** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/209?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#203** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/203?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#202** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/202?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#201** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/201?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#186** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/186?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#185** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/185?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#183** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/183?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#182** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/182?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#181** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/181?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#180** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/180?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#178** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 1 other dependent PR ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#206** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/206?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#205** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/205?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

**Comment by felixmusil** [comment]
In `Ablation: is the compensation necessary in float32?` you don't have any effect of the compensation on the loading but I think it would be a different story for the Heat of Adsorption. Would make a stronger case for the change:smile:

File: test/core/test_kahan.py [hunk 1]
```diff
@@ -78,7 +92,7 @@ def test_difference_recovers_small_increment(self):
         big = jnp.float32(1e5)
         delta = jnp.float32(1e-2)
         old = KahanSummand.init(big)
-        for _ in range(50):
+        for _ in range(5):
```
Comment (felixmusil) [inline test/core/test_kahan.py:103]: why this change?

File: test/core/test_kahan.py [hunk 2]
```diff
@@ -46,6 +46,20 @@ def test_add_two_summands(self):
         b = KahanSummand.init(jnp.array(2.0)) + jnp.array(0.25)
```
Comment (felixmusil) [inline test/core/test_kahan.py:51]: <p class='graphite__hidden'><i>[Re: lines +23 to +24]</i></p>
can you add a test accumulating values that could be out of a gcmc calculation (like computing the HoA so for sure the naive fp32 accumulation would fail) to ensure there is a quick test flagging a regression on this front?
<p class='graphite__hidden'>See this comment inline on <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/267?utm_source=unchanged-line-comment">Graphite</a>.</p>

**Review by felixmusil — CHANGES_REQUESTED** [review]
rather than loading please use the heat of adsorption to showcase the importance of this feature

**Review by felixmusil — APPROVED** [review]

**Comment by n-gao** [comment]
### Merge activity

* **Aug 17, 11:50 AM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/267).
